### ISSUE TITLE
[Bug]: TableSelectAll checkbox checked=true when it's not.

### ISSUE THREAD

**Issue by edumorlom** [carbon-design-system/carbon#15386]
### Package

@carbon/react

### Browser

Chrome, Safari, Firefox, Edge

### Package version

1.42.1

### React version

18.2.0

### Description

├── @carbon/icons-react@11.30.1
├── @carbon/react@1.42.1
├── @types/carbon__icons-react@11.29.0
├── react@18.2.0

### Reproduction/example

none

### Steps to reproduce

<img width="121" alt="Screenshot 2023-12-13 at 12 36 46 PM" src="https://github.com/carbon-design-system/carbon/assets/11444894/78c481ef-e9d4-4cf6-a435-8c44e1d322d7">

Going from `semi-checked` to `unchecked` results on a e.target.checked of `true` even though it should be `false` because the checkbox is unchecked.

```javascript
<TableSelectAll
{...getSelectionProps()}
onSelect={(e) => {
  console.log(e.target.checked)
}}
``` >

### Suggested Severity

None

### Application/PAL

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

**Comment by kuri-sun** [comment]
Hello everyone!!
I took a look at the issue and I found that the code we intentionally set `checked=false` when the TableSelectAll has not selected all rows. (at line 50 in InlineCheckBox.js)
However, I believe that this behavior is correct in terms of the accessibility of this component. Here is why I think so.

checkbox HTML element's accessibility is when a user hits SPACE key to change its state. `checked` represents the checkbox element's internal state. If the user hits the SPACE key to toggle the checkbox state from "indeterminate" to "unselected"(`checked=false`). I suppose "indeterminate"'s `checked` state should be `true`, since once the user hits SPACE key, it is going to be the "unselected" state.

I would like the maintainers to confirm that my assumption is correct or not. If this is not the case, I would love to fix this one!
Thank you so much! ✊

**Comment by edumorlom** [comment]
My issue is that even when the state is completetly unchecked, it shows as checked.

As you can see on the screenshot, the checkbox is completetly OFF but its state is checked=true.


<img width="1736" alt="Screenshot 2023-12-22 at 2 33 47 PM" src="https://github.com/carbon-design-system/carbon/assets/11444894/dab929ec-c3d2-4feb-bc32-988cbb911d1b">

**Comment by kuri-sun** [comment]
Hi @edumorlom! Thank you for the reply.
I completely misunderstood the issue, I am going to take a look at it. I will reply ASAP.
Thanks! ✌️

**Comment by kuri-sun** [comment]
I see. I will take a look and reply to you ASAP! thanks!

**Comment by kuri-sun** [comment]
Hi @edumorlom,
The issue was the `checked` state update was not reflected on the `e.target.checked` value inside `onSelect`. Since we rely on the useEffect to update the `checked` state when the `indeterminate` state has changed.

https://github.com/carbon-design-system/carbon/blob/d177d2685c5767abf7e9798359ebc7b47b13e313/packages/react/src/components/InlineCheckbox/InlineCheckbox.js#L50-L58

I think the solution is invoking this `indeterminate` state update process in `onClick` on <input> HTML element instead of using useEffect.
I will create a PR for this later on. Thank you! ❤️

**Comment by kuri-sun** [comment]
Hi @edumorlom and everyone.
I found the problem cause and its solution so I am sharing them.

(Cause)
As I stated before, we change the `checked` state of the <input> HTML element but since that process depends on useEffect, we can not see the changed value inside the `onClick` function(client side: onSelect) `event` object. So, we need to migrate this updating process to the actual `onClick` function so that users can see the correct `checked` state value associated with `event`.
https://github.com/carbon-design-system/carbon/blob/d177d2685c5767abf7e9798359ebc7b47b13e313/packages/react/src/components/InlineCheckbox/InlineCheckbox.js#L50-L58

(Solution)
Created a new function called `onClickCheckBoxInput` that takes care of this update. Accordingly updated the `onClick` prop in the `inputProps`.
